### DISCUSSION TITLE
GitHub App cannot patch repo visibility in org with repo creation privilege disabled

### DISCUSSION THREAD

**Discussion by jeffwilcox** [community/community#24681] [Enterprise]
<p>Hi,</p>
<p>I wanted to report a painful issue we’re having trying to move to use a GitHub App for common operations. This seems to be an unexpected or invalid set of logic related to the ‘member repository permissions’ set for organizations on GitHub. This is impacting all of our GitHub organizations, including our Enterprise Cloud orgs.</p>
<p><strong>Summary</strong></p>
<ul>
<li>Given a GitHub App with tons of permission (including read/write repository administration, organization administration, etc.)…</li>
<li>The server-to-server GitHub App call to PATCH an existing repository with the body {private:false}, taking a repo public, returns HTTP 422: “Visibility can’t be changed by this user”</li>
<li>This only happens to the GitHub App when the organization’s member privileges are set to not allow repository creation, even if the separate and independent privilege “allow members to change repository visibilities for this organization” is enabled</li>
</ul>
<p><strong>Expected</strong></p>
<p>A PATCH to a repository that is private, to change private to false, succeeds when called by a GitHub App that has repository admin permissions in an organization that has selected the privilege “allow members to change repository visibilities for this organization”</p>
<p><strong>Actual</strong></p>
<p>The PATCH fails with HTTP 422, stating “Visibility can’t be changed by this user”, and pointing to a documentation page about GitHub pricing (although this is a GitHub Enterprise Cloud organization). The repository remains private.</p>
<p><strong>Confirmed API workaround</strong></p>
<p>Change the ‘repository creation’ member privilege restriction to ‘allow private and public’. This does not work with our organization’s security and compliance policy that repos not be created directly on GitHub.</p>
<p><strong>Diagnostic information</strong></p>
<p>(have reproduced with many apps and installs beyond this)</p>
<ul>
<li>GitHub App ID: 41899</li>
<li>GitHub API request ID: 2825:67D5:16B4F9:1B2455:5D93A51C</li>
<li>REST API endpoint: PATCH <a href="https://api.github.com/repos/Azure-Samples/iot-edge-hmi-module" rel="nofollow noopener">https://api.github.com/repos/Azure-Samples/iot-edge-hmi-module</a>
</li>
<li>Body: {“private”:”false”}</li>
<li>Response: HTTP 422</li>
<li>Response body: “{“message”:“Visibility can’t be changed by this user.”,“documentation_url”:“<a href="https://github.com/pricing" rel="nofollow noopener">https://github.com/pricing</a>”}”</li>
</ul>
<p><strong>Customer impact</strong></p>
<p>Cannot use GitHub Apps to replace older workflow. Documented behavior of member privileges is not accurate and causes pain, blocking scenarios.</p>
<p>Painful mitigation – do not use GitHub Apps:</p>
<p>At this time we are forced to abandon using GitHub Apps for some scenarios and will have to continue using org owner personal access tokens or OAuth tokens for authorized users who are org owners.</p>
<p>Confusion around whether GitHub Apps are people/users or not… this is a server-to-server call, but the error message implies it’s a “user”</p>
<p><strong>Set of independent member privileges set for the Azure-Samples org</strong></p>
<p>Member Privileges:</p>
<p>                Repository creation:      </p>
<p>                                Disabled</p>
<p>Admin repository permissions:</p>
<p>                Repository visibility change:</p>
<p>                                [x] Allow members to change repository visibilities for this organization</p>
<p>                                                If enabled, members with admin permissions for the repository will be able</p>
<p>                                                to change repository visibility from public to private. If disabled, only</p>
<p>                                                organization owners can change repository visibilities.</p>
<p>Thanks,</p>
<p>Jeff</p>

**Comment by jeffwilcox** [comment, marked as answer]
<p>FYI, GitHub (thanks) fixed this after Universe.  Thanks API and platform team!</p>

**Reply by svalevka** [reply]
Hi Jeff
Having exactly the same issue.
Is there anything that needs to be done on Github App side?

**Comment by gauntface** [comment]
<p>I’ve just run into this for an org where the “Allow members to change repository visibility for this organization” is disabled.</p>
<p>In this case, I’d still like the app to be able to change visibility given it has Administration permission.</p>
<p>It seems unusual to treat a GitHub App as a member of the org.</p>
